README.md
# synthesis: Synthetic data generator

Generate synthetic time series from commonly used statistical models, including linear, nonlinear and chaotic systems. Applications to testing methods can be found in Jiang, Z., Sharma, A., & Johnson, F. (2019) \doi{10.1016/j.advwatres.2019.103430} and Jiang, Z., Sharma, A., & Johnson, F. (2020) \doi{10.1029/2019WR026962} associated with an open-source tool by Jiang, Z., Rashid, M. M., Johnson, F., & Sharma, A. (2020) \doi{10.1016/j.envsoft.2020.104907}.

It can be used for variable selection, prediction, and classification and clustering problem generation. 

## Requirements
<pre>
Dependencies:
  stats, MASS

Suggest:
  testthat, devtools
</pre>

## Installation

You can install the package via devtools from [GitHub](https://github.com/) with:

``` r
devtools::install_github("zejiang-unsw/synthesis")
```

or via CRAN with: 

``` r
install.packages("synthesis")
```

## Citation

Jiang, Z., Rashid, M. M., Johnson, F., & Sharma, A. (2020). A wavelet-based tool to modulate variance in predictors: An application to predicting drought anomalies. Environmental modelling & software, 135, 104907.

Jiang, Z., Sharma, A., & Johnson, F. (2020). Refining Predictor Spectral Representation Using Wavelet Theory for Improved Natural System Modeling. Water Resources Research, 56(3), e2019WR026962.

Jiang, Z., Sharma, A., & Johnson, F. (2019). Assessing the sensitivity of hydro-climatological change detection methods to model uncertainty and bias. Advances in Water Resources, 134, 103430.

Galelli, S., Humphrey, G. B., Maier, H. R., Castelletti, A., Dandy, G. C., & Gibbs, M. S. (2014). An evaluation framework for input variable selection algorithms for environmental data-driven models. Environmental modelling & software, 62, 33-51.

Sharma, A. (2000). Seasonal to interannual rainfall probabilistic forecasts for improved water supply management: Part 1 - A strategy for system predictor identification. Journal of Hydrology, 239(1), 232-239. 



NEWS.md
# synthesis 1.2.3
* Included applications to testing methods (e.g. WASP and NPRED)

# synthesis 1.2.2
* Included the affine error model.
* Included the Brownian motion models.
* Included the build-up and wash-off water quality model.
* Updated the vignette.

# synthesis 1.2.1
* Added a vignette to the package.
* Included the state-space model.

# synthesis 1.2.0
* Added a `NEWS.md` file to track changes to the package.
* Version submitted to CRAN. 


R/data_gen_BM.R
#' Generate a time series of Brownian motion.
#'
#' This function generates a time series of one dimension Brownian motion.
#'
#' @param x0 the start value of x, with the default value 0
#' @param w0 the start value of w, with the default value 0
#' @param time the temporal interval at which the system will be generated. Default seq(0,by=0.01,len=101).
#' @param do.plot a logical value. If TRUE (default value), a plot of the generated system is shown.
#'
#' @export
#'
#' @references Yanping Chen, http://cos.name/wp-content/uploads/2008/12/stochastic-differential-equation-with-r.pdf
#'
#' @examples
#' set.seed(123)
#' x <- data.gen.bm()
data.gen.bm <- function(x0=0, w0=0, time=seq(0, by=0.01, length.out = 101), do.plot=TRUE)
{
  delta <- diff(time)[1]
  n <- length(time)
  W <- numeric(n)
  t0 <- time[1]

  W[1] <- w0
  for(i in 2:n) W[i] <- W[i-1] + rnorm(1) * sqrt(delta)
  x <- ts(W, start=t0, deltat = delta)

  # plotting
  if (do.plot) {
    title = paste("Brownian motion\n", "x0 = ", x0, "\u0394t = ", delta)
    plot(x, xlab = "t", ylab = "x", main = title, type = "l")
  }

  return(x)
}


#' Generate a time series of geometric Brownian motion.
#'
#' This function generates a a time series of one dimension geometric Brownian motion.
#'
#' @param x0 the start value of x, with the default value 10
#' @param w0 the start value of w, with the default value 0
#' @param mu the interest/drifting rate, with the default value 1.
#' @param sigma the diffusion coefficient, with the default value 0.5.
#' @param time the temporal interval at which the system will be generated. Default seq(0,by=0.01,len=101).
#' @param do.plot a logical value. If TRUE (default value), a plot of the generated system is shown.
#'
#' @export
#'
#' @references Yanping Chen, http://cos.name/wp-content/uploads/2008/12/stochastic-differential-equation-with-r.pdf
#'
#' @examples
#' set.seed(123)
#' x <- data.gen.gbm()
data.gen.gbm <- function(x0=10, w0=0, mu=1, sigma=0.5, time=seq(0, by=0.01, length.out = 101), do.plot=TRUE)
{
  delta <- diff(time)[1]
  n <- length(time)
  W <- numeric(n)
  t0 <- time[1]

  W[1] <- w0
  for(i in 2:n) W[i] <- W[i-1] + rnorm(1) * sqrt(delta)
  S <- x0 * exp((mu-sigma^2/2)*(time-t0) + sigma*(W-W[1]))
  x <- ts(S, start=t0, deltat = delta)

  # plotting
  if (do.plot) {
    title = paste("Geometric Brownian motion\n", "x0 = ", x0, "\u0394t = ", delta,
                                                 "\u03BC = ", mu, "\u03C3 = ", sigma)
    plot(x, xlab = "t", ylab = "x", main = title, type = "l")
  }

  return(x)
}

#' Generate a time series of fractional Brownian motion.
#'
#' This function generates a a time series of one dimension fractional Brownian motion.
#'
#' @param hurst the hurst index, with the default value 0.95, ranging from [0,1].
#' @param time the temporal interval at which the system will be generated. Default seq(0,by=0.01,len=1000).
#' @param do.plot a logical value. If TRUE (default value), a plot of the generated system is shown.
#'
#' @importFrom stats fft
#' @export
#'
#' @references
#' Zdravko Botev (2020). Fractional Brownian motion generator (https://www.mathworks.com/matlabcentral/fileexchange/38935-fractional-brownian-motion-generator), MATLAB Central File Exchange. Retrieved August 17, 2020.
#'
#' Kroese, D. P., & Botev, Z. I. (2015). Spatial Process Simulation. In Stochastic Geometry, Spatial Statistics and Random Fields(pp. 369-404) Springer International Publishing, DOI: 10.1007/978-3-319-10064-7_12
#'
#' @examples
#' set.seed(123)
#' x <- data.gen.fbm()
data.gen.fbm <- function(hurst=0.95, time=seq(0, by=0.01, length.out=1000), do.plot=TRUE)
{
  # Catch errors in inputs
  if(hurst<0|hurst>1) stop('Error: Hurst parameter must be between 0 and 1')

  delta <- diff(time)[1]
  n <- length(time)
  r <- numeric(n+1)

  r[1] <- 1
  for(k in 1:n) r[k+1] <- 0.5 * ((k+1)^(2*hurst) - 2*k^(2*hurst) + (k-1)^(2*hurst))
  r <- c(r, r[seq(length(r)-1, 2)])
  lambda <- Re((fft(r)) / (2*n)) # eigenvalues
  W <- fft(sqrt(lambda) * (rnorm(2*n) + rnorm(2*n)*1i))
  W <- n^(-hurst) * cumsum(Re(W[1:n])) # rescale
  W=(time[n]^hurst)*W # scale for final time T

  x <- ts(W, start=time[1], deltat = delta)

  # plotting
  if (do.plot) {
    title = paste("Fractional Brownian motion\n", "hurst = ", hurst, "\u0394t = ", delta)
    plot(x, xlab = "t", ylab = "x", main = title, type = "l")
  }

  return(x)
}


R/data_gen_SS.R
#' Linear Gaussian state-space model
#'
#' @description
#' Generates data from a specific linear Gaussian state space model of the form
#' \eqn{ x_{t} = \phi x_{t-1} + \sigma_v v_t } and \eqn{ y_t = x_t +
#' \sigma_e e_t }, where \eqn{v_t} and \eqn{e_t} denote independent standard
#' Gaussian random variables, i.e. \eqn{N(0,1)}.
#'
#' @param theta The parameters \eqn{\theta=\{\phi,\sigma_v,\sigma_e\}} of the LGSS model.
#' @param nobs  The data length to be generated.
#' @param start A numeric value indicating the starting value for the time series.
#' If the starting point is not specified, it is generated randomly.
#' @param do.plot Logical value. If TRUE (default value), a plot of the generated LGSS system is shown.
#'
#' @return A list of two variables, state and response.
#' @export
#'
#' @references #Dahlin, J. & Schon, T. B. "Getting Started with Particle Metropolis-Hastings for Inference in Nonlinear Dynamical Models."
#' Journal of Statistical Software, Code Snippets, 88(2): 1--41, 2019.
#'
#' @examples
#' data.LGSS <- data.gen.LGSS(theta=c(0.75,1.00,0.10), nobs=500, start=0)

data.gen.LGSS <- function(theta, nobs, start=runif(n = 1, min = -1, max = 1), do.plot=TRUE)
{
  phi <- theta[1]
  sigmav <- theta[2]
  sigmae <- theta[3]

  state <- matrix(0, nrow = nobs + 1, ncol = 1)
  obs <- matrix(0, nrow = nobs + 1, ncol = 1)

  state[1] <- start
  obs[1] <- NA

  for (t in 2:(nobs + 1)) {
    state[t] <- phi * state[t - 1] + sigmav * rnorm(1)
    obs[t] <- state[t] + sigmae * rnorm(1)
  }

  # plotting
  if (do.plot) {
    title = paste("Linear Gaussian state-space model")
    #plot(1:nobs, x, xlab = "n", ylab = "x[n]", main = title, type = "l")
    plot.ts(cbind(state,obs), main = title, xlab=NA)
  }

  list(x = state, y = obs)
}

#' Stochastic Volatility model


R/data_gen_SW.R
#' Generate predictor and response data: Sinusoidal model
#'
#' @param nobs    The data length to be generated.
#' @param freq    The frequencies in the generated response. Default freq=50.
#' @param A       The amplitude of the sinusoidal series
#' @param phi     The phase of the sinusoidal series
#' @param mu      The mean of Gaussian noise in the variable.
#' @param sd      The standard deviation of Gaussian noise in the variable.
#'
#' @export
#' @return A list of time and x.
#'
#' @references Shumway, R. H., & Stoffer, D. S. (2011). Characteristics of Time Series. In D. S. Stoffer (Ed.), Time series analysis and its applications (pp. 8-14). New York : Springer.
#'
#' @examples
#' ### Sinusoidal model
#' delta <- 1/12 # sampling rate, assuming monthly
#' period.max<- 2^5
#'
#' N = 6*period.max/delta
#' scales<- 2^(0:5)[c(2,6)] #pick two scales
#' scales
#'
#' ### scale, period, and frequency
#' # freq=1/T; T=s/delta so freq = delta/s
#' # since t is t is within 0-1, so freq need to have a factor of N.
#'
#' x1 <- NULL
#' for(i in scales){ # i is the scale
#'
#'   tmp <- synthesis::data.gen.SW(nobs=N, freq = 1/i*N*delta, A = 1, phi = 0, mu=0, sd = 0)$x
#'
#'   x1 <- cbind(x1, tmp)
#' }
#' x <- rowSums(data.frame(x1))
#' plot.ts(cbind(x1,x), type = 'l')

data.gen.SW<-function(nobs=500,freq=50,A=2,phi=pi,mu=0,sd=1)
{

  t <- seq(0,1,length.out = nobs)

  x <- A*cos(2*pi*freq*t + phi) + rnorm(nobs,mean=mu,sd=sd)

  return(list(t=t, x=x))
}



R/data_gen_Chaotic.R
#' Rössler system
#' @description
#' Generates a 3-dimensional time series using the Rossler equations.
#' @details
#' The Rössler system is a system of ordinary differential equations defined as:
#' \deqn{\dot{x} = -(y + z)}{dx/dt = -(y + z)}
#' \deqn{\dot{y} = x+a \cdot y}{dy/dt = x + a*y}
#' \deqn{\dot{z} = b + z*(x-w)}{dz/dt = b + z*(x-w)}
#' The default selection for the system parameters (\emph{a} = 0.2, \emph{b} = 0.2, \emph{w} = 5.7) is known to
#' produce a deterministic chaotic time series. However, the values a = 0.1, b = 0.1, and c = 14 are more commonly used.
#' These Rössler equations are simpler than those Lorenz used since only one nonlinear term appears (the product xz in the third equation).
#'
#' Here, a = b = 0.1 and c changes. The bifurcation diagram reveals that low values of c are periodic,
#' but quickly become chaotic as c increases. This pattern repeats itself as c increases ---
#' there are sections of periodicity interspersed with periods of chaos,
#' and the trend is towards higher-period orbits as c increases.
#' For example, the period one orbit only appears for values of c around 4
#' and is never found again in the bifurcation diagram. The same phenomenon is seen with period three;
#' until c = 12, period three orbits can be found, but thereafter, they do not appear.
#'
#' @param start A 3-dimensional numeric vector indicating the starting point for the time series.
#' Default: c(-2, -10, 0.2).
#' @param a The \emph{a} parameter. Default: 0.2.
#' @param b The \emph{b} parameter. Default: 0.2.
#' @param w The \emph{w} parameter. Default: 5.7.
#' @param time The temporal interval at which the system will be generated.
#' Default: time=seq(0,50,by=0.01) or time = seq(0,by=0.01,length.out = 1000)
#' @param s The level of noise, default 0.
#'
#' @return A list with four vectors named \emph{time}, \emph{x}, \emph{y}
#' and \emph{z} containing the time, the x-components, the
#' y-components and the z-components of the Rössler system, respectively.
#' @export
#' @importFrom graphics plot
#' @importFrom stats plot.ts runif ts
#'
#' @note Some initial values may lead to an unstable system that will tend to infinity.
#' @references Rössler, O. E. 1976. An equation for continuous chaos. Physics Letters A, 57, 397-398.
#' @references Constantino A. Garcia (2019). nonlinearTseries: Nonlinear Time Series Analysis. R package version 0.2.7. https://CRAN.R-project.org/package=nonlinearTseries
#' @references wikipedia https://en.wikipedia.org/wiki/R%C3%B6ssler_attractor
#'
#' @examples
#' ###synthetic example - Rössler
#'
#' ts.r <- data.gen.Rossler(a = 0.1, b = 0.1, w = 8.7, start = c(-2, -10, 0.2),
#'                          time = seq(0, by=0.05, length.out = 10000))
#'
#' oldpar <- par(no.readonly = TRUE)
#' par(mfrow=c(1,1), ps=12, cex.lab=1.5)
#' plot.ts(cbind(ts.r$x,ts.r$y,ts.r$z), col=c('black','red','blue'))
#'
#' par(mfrow=c(1,2), ps=12, cex.lab=1.5)
#' plot(ts.r$x,ts.r$y, xlab='x',ylab = 'y', type = 'l')
#' plot(ts.r$x,ts.r$z, xlab='x',ylab = 'z', type = 'l')
#' par(oldpar)
#'
#' \donttest{
#' # Application to testing variance transformation method in:
#' # Jiang, Z., Sharma, A., & Johnson, F. (2020) <doi:10.1029/2019WR026962>
#' data <- list(x = ts.r$z, dp = cbind(ts.r$x, ts.r$y))
#' dwt <- WASP::dwt.vt(data, wf="d4", J=7, method="dwt", pad="zero", boundary="periodic")
#'
#' par(mfrow = c(ncol(dwt$dp), 1), mar = c(0, 2.5, 2, 1),
#'     oma = c(2, 1, 0, 0), # move plot to the right and up
#'     mgp = c(1.5, 0.5, 0), # move axis labels closer to axis
#'     pty = "m", bg = "transparent",
#'     ps = 12)
#'
#' # plot(dwt$x, type="l", xlab=NA, ylab="SPI12", col="red")
#' # plot(dwt$x, type="l", xlab=NA, ylab="Rain", col="red")
#' for (i in 1:ncol(dwt$dp)) {
#'   ts.plot(cbind(dwt$dp[, i], dwt$dp.n[, i]),
#'           xlab = NA, ylab = NA,
#'           col = c("black", "blue"), lwd = c(1, 2))
#' }
#' }

data.gen.Rossler <- function(a = 0.2, b = 0.2, w = 5.7, start = c(-2, -10, 0.2),
    time = seq(0, by = 0.05, length.out = 1000), s) {
    params <- c(a, b, w)
    rosslerEquations <- function(coord, t, params) {
        x <- coord[[1]]
        y <- coord[[2]]
        z <- coord[[3]]
        a <- params[[1]]
        b <- params[[2]]
        w <- params[[3]]
        c(-y - z, x + a * y, b + z * (x - w))
    }
    r <- rungeKutta(rosslerEquations, start, time, params)

    # add noise
    if (!missing(s)) {
        r[, 1] <- ts(r[, 1] + rnorm(length(time), mean = 0, sd = s))
        r[, 2] <- ts(r[, 2] + rnorm(length(time), mean = 0, sd = s))
        r[, 3] <- ts(r[, 3] + rnorm(length(time), mean = 0, sd = s))
    }

    list(time = time, x = r[, 1], y = r[, 2], z = r[, 3])
}

#' Lorenz system
#' @description
#' Generates a 3-dimensional time series using the Lorenz equations.
#' @details
#' The Lorenz system is a system of ordinary differential equations defined as:
#' \deqn{\dot{x} = \sigma(y-x)}{dx/dt = sigma*( y - x )}
#' \deqn{\dot{y} = \rho x-y-xz}{dy/dt = rho*x - y - xz}
#' \deqn{\dot{z} = -\beta z + xy}{dz/dt = -beta*z + xy}
#' The default selection for the system parameters (\eqn{\sigma=10, \rho=28, \beta=8/3}{sigma=10, rho=28, beta=8/3}) is known to
#' produce a deterministic chaotic time series.
#'
#' @param start A 3-dimensional numeric vector indicating the starting point for the time series.
#' Default: c(-13, -14, 47).
#' @param sigma The \eqn{\sigma}{sigma} parameter. Default: 10.
#' @param rho The \eqn{\rho}{rho} parameter. Default: 28.
#' @param beta The \eqn{\beta}{beta} parameter. Default: 8/3.
#' @param time The temporal interval at which the system will be generated.
#' Default: time=seq(0,50,by = 0.01).
#' @param s The level of noise, default 0.
#'
#' @note Some initial values may lead to an unstable system that will tend to infinity.
#' @references Constantino A. Garcia (2019). nonlinearTseries: Nonlinear Time Series Analysis. R package version 0.2.7. https://CRAN.R-project.org/package=nonlinearTseries
#'
#' @return A list with four vectors named \emph{time}, \emph{x}, \emph{y}
#' and \emph{z} containing the time, the x-components, the
#' y-components and the z-components of the Lorenz system, respectively.
#' @export
#'
#' @examples
#' ###Synthetic example - Lorenz
#' ts.l <- data.gen.Lorenz(sigma = 10, beta = 8/3, rho = 28, start = c(-13, -14, 47),
#'                         time = seq(0, by=0.05, length.out = 2000))
#'
#' ts.plot(cbind(ts.l$x,ts.l$y,ts.l$z), col=c('black','red','blue'))

data.gen.Lorenz <- function(sigma = 10, beta = 8/3, rho = 28, start = c(-13, -14,
    47), time = seq(0, 50, length.out = 1000), s) {
    params <- c(sigma, beta, rho)
    lorenzEquations <- function(coord, t, params) {
        x <- coord[[1]]
        y <- coord[[2]]
        z <- coord[[3]]
        sigma <- params[[1]]
        beta <- params[[2]]
        rho <- params[[3]]
        c(sigma * (y - x), rho * x - y - x * z, x * y - beta * z)
    }
    l <- rungeKutta(lorenzEquations, start, time, params)

    # add noise
    if (!missing(s)) {
        l[, 1] <- ts(l[, 1] + rnorm(length(time), mean = 0, sd = s))
        l[, 2] <- ts(l[, 2] + rnorm(length(time), mean = 0, sd = s))
        l[, 3] <- ts(l[, 3] + rnorm(length(time), mean = 0, sd = s))
    }

    list(time = time, x = l[, 1], y = l[, 2], z = l[, 3])
}

#' Duffing map
#' @description
#' Generates a 2-dimensional time series using the Duffing map.
#' @details
#' The Duffing map is defined as follows:
#' \deqn{ x_n = y_{n - 1}}{x[n] = y[n - 1]}
#' \deqn{ y_n = -b \cdot x_{n - 1} + a \cdot y_{n - 1} - y_{n - 1}^3}{y[n] = -b*x[n - 1] + a*y[n-1] - y[n-1]^3.}
#' The default selection for both \emph{a} and \emph{b} parameters
#' (\emph{a}=1.4 and \emph{b}=0.3) is known to produce a deterministic chaotic
#' time series.
#'
#' @param nobs Length of the generated time series. Default: 5000 samples.
#' @param a The \emph{a} parameter. Default: 2.75.
#' @param b The \emph{b} parameter. Default: 0.2.
#' @param start A 2-dimensional vector indicating the starting values for the x and y Duffing coordinates.
#' Default: If the starting point is not specified, it is generated randomly.
#' @param s The level of noise, default 0.
#' @param do.plot Logical value. If TRUE (default value), a plot of the generated Duffing system is shown.
#'
#' @note Some initial values may lead to an unstable system that will tend to infinity.
#' @references Constantino A. Garcia (2019). nonlinearTseries: Nonlinear Time Series Analysis. R package version 0.2.7. https://CRAN.R-project.org/package=nonlinearTseries
#'
#' @return A list with two vectors named \emph{x} and \emph{y} containing the
#' x-components and the y-components of the Duffing map, respectively.
#' @export
#'
#' @examples
#' Duffing.map=data.gen.Duffing(nobs = 1000, do.plot=TRUE)

data.gen.Duffing <- function(nobs = 5000, a = 2.75, b = 0.2, start = runif(n = 2,
    min = -0.5, max = 0.5), s, do.plot = TRUE) {
    nwarm <- 500
    n <- nobs + nwarm
    y <- x <- vector(mode = "numeric", length = n)
    x[[1]] <- start[[1]]
    y[[1]] <- start[[2]]

    for (i in 2:n) {
        x[[i]] <- y[[i - 1]]
        y[[i]] <- -b * x[[i - 1]] + a * y[[i - 1]] - y[[i - 1]]^3
    }

    # add noise
    if (!missing(s)) {
        x <- x + rnorm(n, mean = 0, sd = s)
        y <- y + rnorm(n, mean = 0, sd = s)
    }

    x <- x[(nwarm + 1):n]
    y <- y[(nwarm + 1):n]

    # plotting
    if (do.plot) {
        title <- paste("Duffing map\n", "a = ", a, " b = ", b)
        plot(x, y, xlab = "x[n]", ylab = "y[n]", main = title, type = "p")
    }
    list(x = x, y = y)
}

#' Henon map
#' @description
#' Generates a 2-dimensional time series using the Henon map.
#' @details
#' The Henon map is defined as follows:
#' \deqn{ x_n = 1 - a \cdot x_{n - 1}^2 + y_{n - 1}}{x[n] = 1 - a*x[n - 1]^2 + y[n - 1]}
#' \deqn{ y_n = b \cdot x_{n - 1}}{y[n] = b*x[n - 1].}
#' The default selection for both \emph{a} and \emph{b} parameters
#' (\emph{a}=1.4 and \emph{b}=0.3) is known to produce a deterministic chaotic
#' time series.
#'
#' @param nobs Length of the generated time series. Default: 5000 samples.
#' @param a The \emph{a} parameter. Default: 1.4.
#' @param b The \emph{b} parameter. Default: 0.3.
#' @param start A 2-dimensional vector indicating the starting values for the x and y Henon coordinates.
#' Default: If the starting point is not specified, it is generated randomly.
#' @param s The level of noise, default 0.
#' @param do.plot Logical value. If TRUE (default value), a plot of the generated Henon system is shown.
#'
#' @note Some initial values may lead to an unstable system that will tend to infinity.
#' @references Constantino A. Garcia (2019). nonlinearTseries: Nonlinear Time Series Analysis. R package version 0.2.7. https://CRAN.R-project.org/package=nonlinearTseries
#'
#' @return A list with two vectors named \emph{x} and \emph{y} containing the
#' x-components and the y-components of the Henon map, respectively.
#' @export
#'
#' @examples
#' Henon.map=data.gen.Henon(nobs = 1000, do.plot=TRUE)

data.gen.Henon <- function(nobs = 5000, a = 1.4, b = 0.3, start = runif(n = 2, min = -0.5,
    max = 0.5), s, do.plot = TRUE) {
    nwarm <- 500
    n <- nobs + nwarm
    y <- x <- vector(mode = "numeric", length = n)
    x[[1]] <- start[[1]]
    y[[1]] <- start[[2]]

    for (i in 2:n) {
        x[[i]] <- y[[i - 1]] + 1 - a * x[[i - 1]]^2
        y[[i]] <- b * x[[i - 1]]
    }

    # add noise
    if (!missing(s)) {
        x <- x + rnorm(n, mean = 0, sd = s)
        y <- y + rnorm(n, mean = 0, sd = s)
    }

    x <- x[(nwarm + 1):n]
    y <- y[(nwarm + 1):n]

    # plotting
    if (do.plot) {
        title <- paste("Henon map\n", "a = ", a, " b = ", b)
        plot(x, y, xlab = "x[n]", ylab = "y[n]", main = title, type = "p")
    }
    list(x = x, y = y)
}

#' Logistic map
#' @description
#' Generates a time series using the logistic map.
#' @details
#' The logistic map is defined as follows:
#' \deqn{x_n = r  \cdot  x_{n-1}   \cdot  (1 - x_{n-1})}{x[n] = r * x[n-1]  * (1 - x[n-1])}
#'
#' @param nobs  Length of the generated time series. Default: 5000 samples.
#' @param r     The \emph{r} parameter. Default: 4
#' @param start A numeric value indicating the starting value for the time series.
#' If the starting point is not specified, it is generated randomly.
#' @param s The level of noise, default 0.
#' @param do.plot Logical value. If TRUE (default value), a plot of the generated Logistic system is shown.
#'
#' @return A vector of time series.
#' @export
#'
#' @references Constantino A. Garcia (2019). nonlinearTseries: Nonlinear Time Series Analysis. R package version 0.2.7. https://CRAN.R-project.org/package=nonlinearTseries
#'
#' @examples
#' Logistic.map=data.gen.Logistic(nobs = 1000, do.plot=TRUE)

data.gen.Logistic <- function(nobs = 5000, r = 4, start = runif(n = 1, min = 0, max = 1),
    s, do.plot = TRUE) {
    nwarm <- 500
    n <- nobs + nwarm
    x <- vector(mode = "numeric", length = n)
    x[[1]] <- start

    for (i in 2:n) {
        x[[i]] <- r * x[[i - 1]] * (1 - x[[i - 1]])
    }

    # add noise
    if (!missing(s)) {
        x <- x + rnorm(n, mean = 0, sd = s)
    }

    x <- x[(nwarm + 1):n]

    # plotting
    if (do.plot) {
        title <- paste("Logistic map\n", "r = ", r)
        # plot(1:nobs, x, xlab = 'n', ylab = 'x[n]', main = title, type = 'l')
        plot(x[-length(x)], x[-1], xlab = "x[t]", ylab = "x[t+1]", main = title,
            type = "p")
    }

    list(x = x)
}

# Runge-Kutta method for solving differential equations. It is used to generate
# both Lorenz and Rossler systems.
rungeKutta <- function(func, initial.condition, time, params) {
    n.samples <- length(time)
    h <- time[[2]] - time[[1]]
    y <- matrix(ncol = length(initial.condition), nrow = n.samples)
    y[1, ] <- initial.condition
    for (i in 2:n.samples) {
        k1 <- h * func(y[i - 1, ], time[[i - 1]], params)
        k2 <- h * func(y[i - 1, ] + k1/2, time[[i - 1]] + h/2, params)
        k3 <- h * func(y[i - 1, ] + k2/2, time[[i - 1]] + h/2, params)
        k4 <- h * func(y[i - 1, ] + k3, time[[i - 1]] + h, params)

        y[i, ] <- y[i - 1, ] + (k1 + 2 * k2 + 2 * k3 + k4)/6
    }
    y
}


R/data_gen_AR.R
#' Generate predictor and response data from AR1 model.
#'
#' @param nobs The data length to be generated.
#' @param ndim The number of potential predictors (default is 9).
#'
#' @return A list of 2 elements: a vector of response (x), and a matrix of potential predictors (dp) with each column containing one potential predictor.
#' @export
#' @importFrom stats rnorm pnorm
#'
#' @examples
#' # AR1 model from paper with 9 dummy variables
#' data.ar1<-data.gen.ar1(500)
#' plot.ts(cbind(data.ar1$x,data.ar1$dp))
#'
#' # Predictor Identifier
#' NPRED::stepwise.PIC(data.ar1$x, data.ar1$dp)

data.gen.ar1 <- function(nobs, ndim = 9) {
    nwarm <- 500
    n <- nobs + nwarm
    x <- matrix(0, n, 1)
    for (i in 1:nwarm) {
        x[i] <- rnorm(1, mean = 0, sd = 1)
    }
    dp <- matrix(0, (nobs), ndim)
    for (i in (nwarm + 1):n) {
        eps <- rnorm(1, mean = 0, sd = 1)
        x[i] <- 0.9 * x[i - 1] + 0.866 * eps
    }
    for (i in 1:ndim) dp[, i] <- x[(n - i - nobs + 1):(n - i)]
    x <- x[(n - nobs + 1):n]
    data_generated <- list(x = x, dp = dp, true.cpy = c(1))
    return(data_generated)
}

#' Generate predictor and response data from AR4 model.
#'
#' @param nobs The data length to be generated.
#' @param ndim The number of potential predictors (default is 9).
#'
#' @return A list of 2 elements: a vector of response (x), and a matrix of potential predictors (dp) with each column containing one potential predictor.
#' @export
#'
#' @examples
#' # AR4 model from paper with total 9 dimensions
#' data.ar4<-data.gen.ar4(500)
#' plot.ts(cbind(data.ar4$x,data.ar4$dp))
#'
#'\donttest{
#' # Predictor Identifier
#' NPRED::stepwise.PIC(data.ar4$x, data.ar4$dp)
#' }

data.gen.ar4 <- function(nobs, ndim = 9) {
    nwarm <- 500
    n <- nobs + nwarm
    x <- matrix(0, n, 1)
    for (i in 1:nwarm) {
        x[i] <- rnorm(1, mean = 0, sd = 1)
    }
    dp <- matrix(0, (nobs), ndim)
    for (i in (nwarm + 1):n) {
        eps <- rnorm(1, mean = 0, sd = 1)
        x[i] <- 0.6 * x[i - 1] - 0.4 * x[i - 4] + eps
    }
    for (i in 1:ndim) dp[, i] <- x[(n - i - nobs + 1):(n - i)]
    x <- x[(n - nobs + 1):n]
    data_generated <- list(x = x, dp = dp, true.cpy = c(1, 4))
    return(data_generated)
}

#' Generate predictor and response data from AR9 model.
#'
#' @param nobs The data length to be generated.
#' @param ndim The number of potential predictors (default is 9).
#'
#' @return A list of 2 elements: a vector of response (x), and a matrix of potential predictors (dp) with each column containing one potential predictor.
#' @export
#'
#' @examples
#' # AR9 model from paper with total 9 dimensions
#' data.ar9<-data.gen.ar9(500)
#' plot.ts(cbind(data.ar9$x,data.ar9$dp))
#'
#'\donttest{
#' # Predictor Identifier
#' NPRED::stepwise.PIC(data.ar9$x, data.ar9$dp)
#' }

data.gen.ar9 <- function(nobs, ndim = 9) {
    nwarm <- 500
    n <- nobs + nwarm
    x <- matrix(0, n, 1)
    for (i in 1:nwarm) {
        x[i] <- rnorm(1, mean = 0, sd = 1)
    }
    dp <- matrix(0, (nobs), ndim)
    for (i in (nwarm + 1):n) {
        eps <- rnorm(1, mean = 0, sd = 1)
        x[i] <- 0.3 * x[i - 1] - 0.6 * x[i - 4] - 0.5 * x[i - 9] + eps
    }
    for (i in 1:ndim) dp[, i] <- x[(n - i - nobs + 1):(n - i)]
    x <- x[(n - nobs + 1):n]
    data_generated <- list(x = x, dp = dp, true.cpy = c(4, 9, 1))
    return(data_generated)
}

#' Generate predictor and response data from TAR1 model.
#'
#' @param nobs  The data length to be generated.
#' @param ndim  The number of potential predictors (default is 9).
#' @param noise The white noise in the data
#'
#' @return A list of 2 elements: a vector of response (x), and a matrix of potential predictors (dp) with each column containing one potential predictor.
#' @export
#'
#' @references Sharma, A. (2000). Seasonal to interannual rainfall probabilistic forecasts for improved water supply management: Part 1 - A strategy for system predictor identification. Journal of Hydrology, 239(1-4), 232-239.
#'
#' @examples
#' # TAR1 model from paper with total 9 dimensions
#' data.tar1<-data.gen.tar1(500)
#' plot.ts(cbind(data.tar1$x,data.tar1$dp))

data.gen.tar1 <- function(nobs, ndim = 9, noise = 0.1) {
    nwarm1 <- nwarm2 <- 250
    n <- nobs + nwarm1 + nwarm2
    x <- matrix(0, n, 1)
    for (i in 1:nwarm1) {
        x[i] <- rnorm(1, mean = 0, sd = 1)
    }
    dp <- matrix(0, (nobs), ndim)
    for (i in (nwarm1 + 1):n) {
        eps <- rnorm(1, mean = 0, sd = 1)
        xi3 <- x[i - 3]
        if (xi3 <= 0)
            x[i] <- -0.9 * x[i - 3] + noise * eps else x[i] <- 0.4 * x[i - 3] + noise * eps
    }
    for (i in 1:ndim) dp[, i] <- x[(n - i - nobs + 1):(n - i)]
    x <- x[(n - nobs + 1):n]
    data_generated <- list(x = x, dp = dp, true.cpy = c(3))
    return(data_generated)
}

#' Generate predictor and response data from TAR2 model.
#'
#' @param nobs  The data length to be generated.
#' @param ndim  The number of potential predictors (default is 9).
#' @param noise The white noise in the data
#'
#' @return A list of 2 elements: a vector of response (x), and a matrix of potential predictors (dp) with each column containing one potential predictor.
#' @export
#'
#' @references Sharma, A. (2000). Seasonal to interannual rainfall probabilistic forecasts for improved water supply management: Part 1 - A strategy for system predictor identification. Journal of Hydrology, 239(1-4), 232-239.
#'
#' @examples
#' # TAR2 model from paper with total 9 dimensions
#' data.tar2<-data.gen.tar2(500)
#' plot.ts(cbind(data.tar2$x,data.tar2$dp))

data.gen.tar2 <- function(nobs, ndim = 9, noise = 0.1) {
    nwarm1 <- nwarm2 <- 250
    n <- nobs + nwarm1 + nwarm2
    x <- matrix(0, n, 1)
    for (i in 1:nwarm1) {
        x[i] <- rnorm(1, mean = 0, sd = 1)
    }
    dp <- matrix(0, (nobs), ndim)
    for (i in (nwarm1 + 1):n) {
        eps <- rnorm(1, mean = 0, sd = 1)
        xi6 <- x[i - 6]
        if (xi6 <= 0)
            x[i] <- -0.5 * x[i - 6] + 0.5 * x[i - 10] + noise * eps else x[i] <- 0.8 * x[i - 10] + noise * eps
    }
    for (i in 1:ndim) dp[, i] <- x[(n - i - nobs + 1):(n - i)]
    x <- x[(n - nobs + 1):n]
    data_generated <- list(x = x, dp = dp, true.cpy = c(10, 6))
    return(data_generated)
}


